lib/mix/test/mix/tasks/deps_test.exs

```diff
@@ -527,7 +535,8 @@ defmodule Mix.Tasks.DepsTest do
         Mix.Tasks.Deps.Loadpaths.run([])
       end
 
-      assert_received {:mix_shell, :error, [@overriding_msg <> _]}
```

Comment:
Allowing this, creates a inconsistency on the error:
```
iex(2)> 1 <> "b" = "a" 
** (CompileError) iex:2: The left argument of <> operator inside a match should be always a literal binary
    (elixir) lib/kernel.ex:1633: Kernel.check_invalid_argument/2
    (elixir) lib/kernel.ex:1608: Kernel.extract_concatenations/2
    (elixir) expanding macro: Kernel.<>/2
    iex:2: (file)
iex(2)> defmodule B do        
...(2)> @foo 1                
...(2)> def b, do: @foo <> "b" = "ab"
...(2)> end
** (CompileError) iex:4: conflicting type specification for bit field: "binary" and "integer"
    (elixir) src/elixir_bitstring.erl:162: :elixir_bitstring.expand_specs/5
    (elixir) src/elixir_bitstring.erl:35: :elixir_bitstring.expand/7
    (elixir) src/elixir_bitstring.erl:10: :elixir_bitstring.expand/4
    (elixir) expanding macro: Kernel.<>/2
    iex:4: B.b/0
```

Reply:
We cannot check if the left side is a binary inside the `<>` macro because at that moment the arguments have not been expanded yet. For example, the argument could be this `foo() <> "bar" = ...` where `foo()` is a macro that expands to a binary.

We could expand the arguments inside `<>` but we will certainly run into bootstrapping issues.

My suggestion is to check exclusively if `^something` is used on the left side of `<>` and raise only in that case. For the other cases we need to make sure that a good message comes from inside the compiler binary parser. In particular, this is NOT a good error message:

```iex
iex(1)> 1 <> binary = 1
** (CompileError) iex:1: conflicting type specification for bit field: "binary" and "integer"
```

We should be able to say that "you have declared type binary but got argument of type integer: 1".

We can open a separate issue if you don't want to tackle both issues here. :)

Reply:
I didn't check exclusively for the ^ operator because the same problem happens for vars:
```
iex(1)> x <> "b" = "ab"
** (CompileError) iex:1: a binary field without size is only allowed at the end of a binary pattern and never allowed in binary generators
    (elixir) src/elixir_bitstring.erl:88: :elixir_bitstring.expand_specs/5
    (elixir) src/elixir_bitstring.erl:29: :elixir_bitstring.expand/6
    (elixir) src/elixir_bitstring.erl:9: :elixir_bitstring.expand/4
    (elixir) expanding macro: Kernel.<>/2
    iex:1: (file)
```

Although, I agree with you about the problems created. I've changed that to specifically look for the pin operator and the var.

About the types, I added a context to the expansion, but I'm not sure. I think the problem is that the message does perfect sense when using the <<>> operator:
```
iex(1)> <<1::binary, "bar">>
** (CompileError) iex:1: conflicting type specification for bit field: "binary" and "integer"
    (elixir) src/elixir_bitstring.erl:87: :elixir_bitstring.expand_specs/5
    (elixir) src/elixir_bitstring.erl:29: :elixir_bitstring.expand/6
    (elixir) src/elixir_bitstring.erl:12: :elixir_bitstring.expand/4
```

I have the impression that the compiler needs more information to give a good message

Reply:
Maybe this conflicts with the previous idea of letting the errors close with the <> definition